tests/test_experiment.py

```diff
@@ -389,3 +393,135 @@ def impl(*a, **kw):
+    assert time.perf_counter() - now < 1
+
+
+def test_wait_calls_prefered_impl(make_populated_experment, monkeypatch):
+    """Make wait is calling the expected method for checking job statuses.
+    Right now we only have the "polling" impl, but in future this might change
+    to an event based system.
+    """
+    exp = make_populated_experment(1)
+    ((_, (id_, *_)),) = exp._launch_history.group_by_launcher().items()
+    was_called = False
+
+    def mocked_impl(*args, **kwargs):
+        nonlocal was_called
+        was_called = True
+
+    monkeypatch.setattr(exp, "_poll_for_statuses", mocked_impl)
+    exp.wait(id_)
+    assert was_called
+
+
+@pytest.mark.parametrize(
+    "num_polls",
+    [
+        pytest.param(i, id=f"Poll for status {i} times")
+        for i in (1, 5, 10, 20, 100, 1_000)
+    ],
+)
+@pytest.mark.parametrize("verbose", [True, False])
+def test_poll_status_blocks_until_job_is_completed(
+    monkeypatch, make_populated_experment, num_polls, verbose
+):
+    """Make sure that the polling based impl blocks the calling thread. Use
+    varying number of polls to simulate varying lengths of job time for a job
+    to complete.
+
+    Additionally check to make sure that the expected log messages are present
+    """
+    exp = make_populated_experment(1)
+    ((launcher, (id_, *_)),) = exp._launch_history.group_by_launcher().items()
+    (current_status,) = launcher.get_status(id_).values()
+    different_statuses = set(JobStatus) - {current_status}
+    (new_status, *_) = different_statuses
+    mock_log = io.StringIO()
+
+    @dataclasses.dataclass
+    class ChangeStatusAfterNPolls:
+        n: int
+        from_: JobStatus
+        to: JobStatus
+        num_calls: int = dataclasses.field(default=0, init=False)
+
+        def __call__(self, *args, **kwargs):
+            self.num_calls += 1
+            ret_status = self.to if self.num_calls >= self.n else self.from_
+            return (ret_status,)
+
+    mock_get_status = ChangeStatusAfterNPolls(num_polls, current_status, new_status)
+    monkeypatch.setattr(exp, "get_status", mock_get_status)
+    monkeypatch.setattr(
+        "smartsim.experiment.logger.info", lambda s: mock_log.write(f"{s}\n")
+    )
+    final_statuses = exp._poll_for_statuses(
+        [id_],
+        different_statuses,
+        timeout=SynchronousTimeInterval(10),
+        interval=SynchronousTimeInterval(0),
+        verbose=verbose,
+    )
+    assert final_statuses == {id_: new_status}
+
+    expected_log = io.StringIO()
+    expected_log.writelines(
+        f"Job({id_}): Running with status '{current_status.value}'\n"
+        for _ in range(num_polls - 1)
+    )
+    expected_log.write(f"Job({id_}): Finished with status '{new_status.value}'\n")
+    assert mock_get_status.num_calls == num_polls
+    assert mock_log.getvalue() == (expected_log.getvalue() if verbose else "")
+
+
+def test_poll_status_raises_when_called_with_infinite_iter_wait(
+    make_populated_experment,
```

Comment:
typo: change to "make_populated_experiment"

Reply:
🤦